### PR TITLE
OpenBSD Behavior with Multiple DOWN Interfaces + snprintf Downstream Patch

### PR DESCRIPTION
**Pull request by dohnuts**
* Fixing Openbsd Behavior with Multiple DOWN Interface 
move to cleaner libc getiffaddrs :
  * 2002 in glibc
  * uClibc available  https://github.com/hwoarang/uClibc/blob/master-metag/libc/inet/ifaddrs.c 
  * Since 'OpenBSD 2.7'
* Using snprintf as patched downstream in 1.0
* Remove the timeb include that block builds

### CODE REVIEW

**Comment by sgeto** [comment]
So the important thing is that patches don't break existing setups and OSs we already support. Macos in particular could be effected by your fix, but as far as I can tell, everything seems pretty normal there.
I've set up continuous integration a while ago, but it's kind of unfinished. The checks it performs **after** building are quite strict and our master branch isn't currently setup to these standard. Nonetheless building succeeds with your patches on both Linux and macos, which is all that matter.
https://travis-ci.org/sgeto/libnet/builds/529880188

@sam-github thoughts?

**Comment by sgeto** [comment]
@dohnuts 
Could you please remove all references of of the macro `NO_SNPRINTF`?
There is at least one in include/libnet-macros.c. They aren't needed anymore.
Thanks

File: libnet/src/libnet_if_addr.c [hunk 1]
```diff
@@ -76,13 +76,62 @@ libnet_check_iface(libnet_t *l)
         {
             snprintf(l->err_buf, LIBNET_ERRBUF_SIZE, "%s(): %s is down",
                     __func__, l->device);
-	    res = -1;
+           res = -1;
```
Comment (sam-github) [inline libnet/src/libnet_if_addr.c]: why the whitespace change?

File: libnet/src/libnet_if_addr.c [hunk 2]
```diff
@@ -76,13 +76,62 @@ libnet_check_iface(libnet_t *l)
...
+#include <sys/types.h>
+    #ifdef __OpenBSD__ 
+    #include <sys/socket.h> 
+    #endif
+#include <ifaddrs.h>
+
+int
+libnet_ifaddrlist(register struct libnet_ifaddr_list **ipaddrp, char *dev, register char *errbuf)
+{
+    static struct libnet_ifaddr_list ifaddrlist[MAX_IPADDR]; /* REALLY ? */
```
Comment (sam-github) [inline libnet/src/libnet_if_addr.c]: "REALLY?" means what?

File: libnet/src/libnet_if_addr.c [hunk 3]
```diff
@@ -76,13 +76,62 @@ libnet_check_iface(libnet_t *l)
...
+    for (ifa = ifap; ifa; ifa = ifa->ifa_next)
+    {
+        if (ifa->ifa_flags & IFF_LOOPBACK)
+            continue;
+
+        if (ifa->ifa_addr->sa_family == AF_INET )
+        {
+            ifaddrlist[i].device = strdup(ifa->ifa_name);
+            if (ifaddrlist[i].device == NULL) {
+                snprintf(errbuf, LIBNET_ERRBUF_SIZE, "%s(): OOM",__func__);
```
Comment (sam-github) [inline libnet/src/libnet_if_addr.c]: missing space in `,_func_`.

File: libnet/src/libnet_if_addr.c [hunk 4]
```diff
@@ -386,7 +437,7 @@ libnet_select_device(libnet_t *l)
                 "%s(): no network interface found", __func__);
         return (-1);
     }
-	
+
```
Comment (sam-github) [inline libnet/src/libnet_if_addr.c]: are you stripping trailing whitespace? I approve, but its cosmetic, better put in another commit (or leave alone).

File: libnet/src/libnet_if_addr.c [hunk 5]
```diff
@@ -411,7 +462,7 @@ libnet_select_device(libnet_t *l)
             snprintf(l->err_buf, LIBNET_ERRBUF_SIZE,
                     "%s(): can't find interface for IP %s", __func__,
                     l->device);
-	    goto bad;
+            goto bad;
```
Comment (sam-github) [inline libnet/src/libnet_if_addr.c]: unrelated whitespace change

File: libnet/src/libnet_link_bpf.c [hunk 6]
```diff
@@ -61,7 +63,7 @@ libnet_bpf_open(char *err_buf)
      */
     for (i = 0;;i++)
     {
-        sprintf(device, "/dev/bpf%d", i);
+        snprintf(device, 11, "/dev/bpf%d", i);
```
Comment (sam-github) [inline libnet/src/libnet_link_bpf.c]: `11` should be `sizeof(device)`

Comment (sam-github) [reply libnet/src/libnet_link_bpf.c]: also, its a pre-existing bug, but perhaps you should assert if i > 999, since the code won't work.

File: libnet/src/libnet_link_dlpi.c [hunk 7]
```diff
@@ -228,7 +228,7 @@ libnet_open_link(libnet_t *l)
     }
     else
     {
-        sprintf(dname, "%s/%s", DLPI_DEV_PREFIX, l->device);
+        snprintf(dname, MAXPATHLEN,"%s/%s", DLPI_DEV_PREFIX, l->device);
```
Comment (sam-github) [inline libnet/src/libnet_link_dlpi.c]: please put a ` ` after the `,`s

File: libnet/src/libnet_link_nit.c [hunk 8]
```diff
@@ -52,7 +52,8 @@ libnet_open_link_interface(int8_t *device, int8_t *ebuf)
     l->fd = socket(AF_NIT, SOCK_RAW, NITPROTO_RAW);
     if (l->fd < 0)
     {
-        sprintf(ebuf, "socket: %s", strerror(errno));
+                 snprintf(ebuf, LIBNET_ERRBUF_SIZE,
+"socket: %s", strerror(errno));
```
Comment (sam-github) [inline libnet/src/libnet_link_nit.c]: the indentation got messed up here

File: libnet/src/libnet_link_nit.c [hunk 9]
```diff
@@ -61,7 +62,8 @@ libnet_open_link_interface(int8_t *device, int8_t *ebuf)
 
     if (bind(l->fd, (struct sockaddr *)&snit, sizeof(snit)))
     {
-        sprintf(ebuf, "bind: %s: %s", snit.snit_ifname, strerror(errno));
+                 snprintf(ebuf, LIBNET_ERRBUF_SIZE,
+"bind: %s: %s", snit.snit_ifname, strerror(errno));
```
Comment (sam-github) [inline libnet/src/libnet_link_nit.c]: indentation

File: libnet/src/libnet_link_pf.c [hunk 10]
```diff
@@ -49,22 +49,25 @@ libnet_open_link_interface(int8_t *device, int8_t *ebuf)
     l = (struct libnet_link_int *)malloc(sizeof(*l));
     if (l == NULL)
     {
-        sprintf(ebuf, "libnet_open_link_int: %s", strerror(errno));
+                 snprintf(ebuf, LIBNET_ERRBUF_SIZE,
+"libnet_open_link_int: %s", strerror(errno));
```
Comment (sam-github) [inline libnet/src/libnet_link_pf.c]: indentation

File: libnet/src/libnet_link_pf.c [hunk 11]
```diff
@@ -49,22 +49,25 @@ libnet_open_link_interface(int8_t *device, int8_t *ebuf)
...
         return (0);
     }
     memset(l, 0, sizeof(*l));
     l->fd = pfopen(device, O_RDWR);
     if (l->fd < 0)
     {
-        sprintf(ebuf, "pf open: %s: %s\n\your system may not be properly configured; see \"man packetfilter(4)\"",
+                 snprintf(ebuf, LIBNET_ERRBUF_SIZE,
+"pf open: %s: %s\n\your system may not be properly configured; see \"man packetfilter(4)\"",
```
Comment (sam-github) [inline libnet/src/libnet_link_pf.c]: indentation, and existing bug: `\y`, since you are touching the line, could you fix that, please?

File: libnet/src/libnet_link_pf.c [hunk 12]
```diff
@@ -49,22 +49,25 @@ libnet_open_link_interface(int8_t *device, int8_t *ebuf)
...
             device, strerror(errno));
         goto bad;
     }
 
     enmode = ENTSTAMP|ENBATCH|ENNONEXCL;
     if (ioctl(l->fd, EIOCMBIS, (caddr_t)&enmode) < 0)
     {
-        sprintf(ebuf, "EIOCMBIS: %s", strerror(errno));
+                 snprintf(ebuf, LIBNET_ERRBUF_SIZE,
+"EIOCMBIS: %s", strerror(errno));
```
Comment (sam-github) [inline libnet/src/libnet_link_pf.c]: I'll stop commenting, but pls check for damaged indentation

**Comment by dohnuts** [comment]
> @dohnuts
> Could you please remove all references of of the macro `NO_SNPRINTF`?
> There is at least one in include/libnet-macros.c. They aren't needed anymore.
> Thanks

This require a change in automake , as the base reference come form configure

*cygwin*)

$as_echo "#define LIBNET_BSDISH_OS 1" >>confdefs.h


$as_echo "#define NO_SNPRINTF 1" >>confdefs.h


i ll advice to separate this as a different problem

**Comment by dohnuts** [comment]
Hello @sam-github ,

I guess i edited the pull request ( i think it s should be satisfactory )

The breaking indentation was no intended of course , 🍁Sorry 🍁

**Comment by sgeto** [comment]
`NO_SNPRINTF` isn't needed. Cygwin once needed this but it came a long way since then.  Drop them from the both the header and configure.ac

**Comment by dohnuts** [comment]
NO_SNPRINTF  removed.

**Comment by dohnuts** [comment]
https://github.com/dohnuts/libnet/commits/style may interest you . It build here.

Note i tried to check cygwin build

and 

/usr/lib
$ objdump.exe -t ./packet.lib  | grep -i PacketSetMode
[  3](sec  1)(fl 0x00)(ty   0)(scl   2) (nx 0) 0x0000000000000000 __imp_PacketSetMode
[  5](sec  3)(fl 0x00)(ty   0)(scl   2) (nx 0) 0x0000000000000000 PacketSetMode

```
checking link-layer packet interface type... found link layer win32 wpcap
scanning available packet construction modules: 802.1q 802.1x 802.2 802.3 arp bgp cdp data dhcp dns ethernet fddi gre hsrp icmp igmp ip ipsec isl link mpls ntp ospf rip rpc sebek snmp stp tcp token_ring udp vrrp
checking for PacketSetMode in -lpacket... no
configure: error: packet lib not found.
```


Which is completly out of scope

I hope you can review / accept #81 

Best

**Comment by dohnuts** [comment]
@sam-github bump

**Comment by dohnuts** [comment]
¯\_(ツ)_/¯

**Comment by dohnuts** [comment]
@sam-github bump

**Comment by troglobit** [comment]
Looks good to me, and I've managed to merge this manually with the current master and builds on my Ubuntu machine without any issues.   I say merge this, unless you have any more comments @sam-github ?

**Comment by sam-github** [comment]
Code looks good. Myself, I'd rewrite the branch history to squash the fixup commit and resolve conflicts.   How would you like to land it @troglobit ?

**Comment by troglobit** [comment]
Well, from the command line there were no conflicts. Git is great at merging branches and I would've been happy and normally stopped there. For this case an interactive rebase to touch-up the commit messages turned out OK, but on any signs of trouble (new conflicts etc.) I would back out and keep the commits as-is just to save a lot of time/pain for me.  The real downside to rebase, however, is that we lose any reference to the commit hashes we've discussed in a PR. So for the audit trail alone I'd keep the commits and do a merge commit.

**Comment by troglobit** [comment]
There; rebased, fixed, and pushed to master! \o/
